### PR TITLE
#1 documentation

### PR DESCRIPTION
**Pull request by smirnovV**

### CODE REVIEW

File: src/main/java/ru/smirnovv/FindMax.java [hunk 1]
```diff
@@ -4,21 +4,31 @@
...
 class FindMax {
+    /**
+     * The number of elements for print.
+     */
     @Parameters(index = "0",
             description = "Number of maximums (must be positive).")
     private Integer numberOfMaximums;
 
+    /**
+     * @param args the arguments for program
```
Comment (01sas) [inline src/main/java/ru/smirnovv/FindMax.java]: Please add the description for the method itself, not only for arguments, like "The entry point.".

Comment (01sas) [inline src/main/java/ru/smirnovv/FindMax.java]: Please put a dot the end of the line.

File: src/main/java/ru/smirnovv/MaxBoundedTreeSet.java [hunk 2]
```diff
@@ -5,17 +5,44 @@
...
+    /**
+     * The maximum number of elements in the MaxBoundedTreeSet.
+     */
     private final long maxSize;
 
+    /**
+     * Constructs an empty MaxBoundedTreeSet with the specified
+     * maximum number of elements.
+     *
+     * @param maxSize the maximum number of elements
```
Comment (01sas) [inline src/main/java/ru/smirnovv/MaxBoundedTreeSet.java]: Please put a dot at the end of the line.

File: src/main/java/ru/smirnovv/FindMax.java [hunk 3]
```diff
@@ -4,21 +4,31 @@
 import picocli.CommandLine;
 
 import java.io.BufferedReader;
+import java.io.IOException;
 import java.io.InputStreamReader;
 
 import static java.lang.System.err;
 import static java.lang.System.out;
 import static picocli.CommandLine.*;
```
Comment (01sas) [inline src/main/java/ru/smirnovv/FindMax.java]: Plese, avoid using of wildcard import.

File: src/main/java/ru/smirnovv/FindMax.java [hunk 4]
```diff
@@ -4,21 +4,31 @@
...
 import static ru.smirnovv.MyCollectors.maxN;
 
+/**
+ * Program for find the given number of maximums.
```
Comment (01sas) [inline src/main/java/ru/smirnovv/FindMax.java]: "Program that finds the given number of maximums."

File: src/main/java/ru/smirnovv/FindMax.java [hunk 5]
```diff
@@ -4,21 +4,31 @@
...
+ */
 @Command(description = "Finds the given number of maximums.",
         name = "FindMax", mixinStandardHelpOptions = true,
         version = "FindMax 0.1.0")
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/FindMax.java]: Would be great to replace the name and version attributes with "${project.artifactId}" and "${project.version}" accordingly and to enable filtering of java source files in Maven that will place the actual artifactId and version into the attribute values. Another way is just to add the version info into the Manifest file and to implement IVersionProvider from Picocli that should extract the version info from the Manifest file in runtime.

File: src/main/java/ru/smirnovv/FindMax.java (hunk 4 above)
Comment (aanbrn) [reply src/main/java/ru/smirnovv/FindMax.java]: Finds the given number of maximums reading lines from the standard input.

File: src/main/java/ru/smirnovv/FindMax.java [hunk 6]
```diff
@@ -4,21 +4,31 @@
...
 class FindMax {
+    /**
+     * The number of elements for print.
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/FindMax.java]: The number of maximums to find.

File: src/main/java/ru/smirnovv/FindMax.java (hunk 1 above)
Comment (aanbrn) [reply src/main/java/ru/smirnovv/FindMax.java]: Yeah, and use "the program arguments" for args.

File: src/main/java/ru/smirnovv/FindMax.java [hunk 7]
```diff
@@ -43,13 +53,19 @@ public static void main(final String[] args) {
         }
     }
 
-    private void run() throws Exception {
-        try (BufferedReader br = new BufferedReader(new InputStreamReader(System.in))) {
+    /**
+     * Method, where read lines and print maximum numbers.
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/FindMax.java]: Executes the program flow reading lines from the standard input, finding maximum numbers in them and printing the result to the standard output.

File: src/main/java/ru/smirnovv/FindMax.java [hunk 8]
```diff
@@ -43,13 +53,19 @@ public static void main(final String[] args) {
...
+     *
+     * @throws IOException if the BufferedReader throws IOException
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/FindMax.java]: if an IO error occurred.

File: src/main/java/ru/smirnovv/MaxBoundedTreeSet.java [hunk 9]
```diff
@@ -5,17 +5,44 @@
 import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkNotNull;
 
+/**
+ * MaxBoundedTreeSet extends the <tt>TreeSet</tt> class.
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MaxBoundedTreeSet.java]: A bounded TreeSet of Long numbers limited by the maximum size and intended to collect the maximum elements.

File: src/main/java/ru/smirnovv/MaxBoundedTreeSet.java [hunk 10]
```diff
@@ -5,17 +5,44 @@
...
+ * This class is limited TreeSet, when elements are sorted in
+ * descending order.
+ *
+ * @see TreeSet
+ */
 public class MaxBoundedTreeSet extends TreeSet<Long> {
+    /**
+     * The maximum number of elements in the MaxBoundedTreeSet.
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MaxBoundedTreeSet.java]: The maximum number of elements in this set.

File: src/main/java/ru/smirnovv/MaxBoundedTreeSet.java [hunk 11]
```diff
@@ -5,17 +5,44 @@
...
+     */
     private final long maxSize;
 
+    /**
+     * Constructs an empty MaxBoundedTreeSet with the specified
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MaxBoundedTreeSet.java]: Constructs an instance with the given maximum size.

File: src/main/java/ru/smirnovv/MaxBoundedTreeSet.java (hunk 2 above)
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MaxBoundedTreeSet.java]: the maximum size.

File: src/main/java/ru/smirnovv/MaxBoundedTreeSet.java [hunk 12]
```diff
@@ -5,17 +5,44 @@
...
+     * maximum number of elements.
+     *
+     * @param maxSize the maximum number of elements
+     * @throws IllegalArgumentException if the specified maximum number
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MaxBoundedTreeSet.java]: if 'maxSize' is not positive.

File: src/main/java/ru/smirnovv/MaxBoundedTreeSet.java [hunk 13]
```diff
@@ -5,17 +5,44 @@
...
+     *                                  of elements is negative or zero
+     */
     public MaxBoundedTreeSet(/*@Nonnegative*/ final int maxSize) {
         checkArgument(maxSize > 0, "Argument 'maxSize' must be positive");
 
         this.maxSize = maxSize;
     }
 
+    /**
+     * Adds the specified element to this set if this element greater
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MaxBoundedTreeSet.java]: Adds a new element to this set if it's not full yet or if the given element is greater than the smallest one in it. When this set is full already and the given element added, it removes the smallest one to keep the maximum size.

File: src/main/java/ru/smirnovv/MaxBoundedTreeSet.java [hunk 14]
```diff
@@ -5,17 +5,44 @@
...
+     * than the smallest element in the MaxBoundedTreeSet or number
+     * of elements less than maxSize.
+     *
+     * @param e element to be added to this set
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MaxBoundedTreeSet.java]: a number to be added.

File: src/main/java/ru/smirnovv/MaxBoundedTreeSet.java [hunk 15]
```diff
@@ -5,17 +5,44 @@
...
+     * @return {@code true} if element added to the MaxBoundedTreeSet
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MaxBoundedTreeSet.java]: whether the given element added to this set or not.

File: src/main/java/ru/smirnovv/MaxBoundedTreeSet.java [hunk 16]
```diff
@@ -5,17 +5,44 @@
...
+     * @throws NullPointerException if the specified element is null
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MaxBoundedTreeSet.java]: if the given number is null.

File: src/main/java/ru/smirnovv/MaxBoundedTreeSet.java [hunk 17]
```diff
@@ -5,17 +5,44 @@
...
+     */
     @Override
-    public boolean add(/*@NotNull*/final Long e) {
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MaxBoundedTreeSet.java]: Let's rename e to n.

File: src/main/java/ru/smirnovv/MyCollectors.java [hunk 18]
```diff
@@ -4,7 +4,27 @@
 
 import static com.google.common.base.Preconditions.checkArgument;
 
-public class MyCollectors {
+/**
+ * This class contains methods, that returns new collectors.
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MyCollectors.java]: A utility class containing factory-methods for own collectors.

File: src/main/java/ru/smirnovv/MyCollectors.java [hunk 19]
```diff
@@ -4,7 +4,27 @@
...
+ */
+public final class MyCollectors {
+    /**
+     * Don't let anyone instantiate this class.
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MyCollectors.java]: Private constructor to block instantiation.

File: src/main/java/ru/smirnovv/MyCollectors.java [hunk 20]
```diff
@@ -4,7 +4,27 @@
...
+     */
+    private MyCollectors() {
+    }
+
+    /**
+     * Return the new collector described by the MaxBoundedTreeSet methods.
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MyCollectors.java]: Returns a collector that accumulates the input elements into a bounded set of maximum ones.

File: src/main/java/ru/smirnovv/MyCollectors.java [hunk 21]
```diff
@@ -4,7 +4,27 @@
...
+     *
+     * @param numberOfMaxes the number of elements for initialization
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MyCollectors.java]: Let's rename numberOfMaxes to numberOfMaximums.

Comment (aanbrn) [reply src/main/java/ru/smirnovv/MyCollectors.java]: the maximum number of maximum elements.

File: src/main/java/ru/smirnovv/MyCollectors.java [hunk 22]
```diff
@@ -4,7 +4,27 @@
...
+     *                      MaxBoundedTreeSet
+     * @return the new {@code collector}
+     * @throws IllegalArgumentException if the specified number of maxes
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MyCollectors.java]: if 'numberOfMaximums' is not positive.

**Review by aanbrn — CHANGES_REQUESTED** [review]
Please take a look at my comments.

**Review by 01sas — APPROVED** [review]

File: src/main/java/ru/smirnovv/FindMax.java [hunk 23]
```diff
@@ -52,4 +76,15 @@ private void run() throws Exception {
...
         }
     }
-}
\ No newline at end of file
+
+    public static class VersionProvider implements CommandLine.IVersionProvider {
+        public String[] getVersion() throws Exception {
+            URLClassLoader cl = (URLClassLoader) FindMax.class.getClassLoader();
+            URL url = cl.findResource("META-INF/MANIFEST.MF");
+            Manifest manifest = new Manifest(url.openStream());
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/FindMax.java]: Probably better to close the opened stream.

File: src/main/java/ru/smirnovv/FindMax.java [hunk 24]
```diff
@@ -52,4 +76,15 @@ private void run() throws Exception {
...
+            Attributes mainAttributes = manifest.getMainAttributes();
+            String implVersion = mainAttributes.getValue("Implementation-Version");
+            return new String[]{implVersion};
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/FindMax.java]: Doesn't look formatted.

File: src/main/java/ru/smirnovv/FindMax.java [hunk 25]
```diff
@@ -52,4 +76,15 @@ private void run() throws Exception {
                     .forEach(out::println);
         }
     }
-}
\ No newline at end of file
+
+    public static class VersionProvider implements CommandLine.IVersionProvider {
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/FindMax.java]: Where is javadoc?

File: src/main/java/ru/smirnovv/MyCollectors.java [hunk 26]
```diff
@@ -4,14 +4,30 @@
...
+    /**
+     * Returns a collector that accumulates the input elements into a bounded set of maximum ones.
+     *
+     * @param numberOfMaximums the maximum number of maximum elements.
+     * @return the new {@code collector}.
+     * @throws IllegalArgumentException if 'numberOfMaximums' is not positive.
+     * @see Collector
+     * @see MaxBoundedTreeSet
+     */
+    public static Collector<Long, MaxBoundedTreeSet, MaxBoundedTreeSet> maxN(/*@Nonnegative*/ final int numberOfMaximums) {
```
Comment (aanbrn) [inline src/main/java/ru/smirnovv/MyCollectors.java]: This line looks quite long and probably needs to be formatted.

**Review by aanbrn — CHANGES_REQUESTED** [review]
See my new comments.

**Review by 01sas — APPROVED** [review]

**Review by aanbrn — APPROVED** [review]

**Pull request closed without merging**